tracing-subscriber/src/filter/layer_filters.rs

```diff
@@ -146,19 +218,45 @@ where
     // almsot certainly impossible...right?
 
     fn register_callsite(&self, metadata: &'static Metadata<'static>) -> Interest {
-        // self.filter.callsite_enabled(metadata)
-        Interest::sometimes()
+        let interest = self.filter.callsite_enabled(metadata);
+
+        // If the filter didn't disable the callsite, allow the inner layer to
+        // register it --- since `register_callsite` is also used for purposes
+        // such as reserving/caching per-callsite data, we want the inner layer
+        // to be able to perform any other registration steps. However, we'll
+        // ignore its `Interest`.
+        if !interest.is_never() {
+            self.layer.register_callsite(metadata);
+        }
+
+        // Add our `Interest` to the current sum of per-layer filter `Interest`s
+        // for this callsite.
+        FILTERING.with(|filtering| filtering.add_interest(interest));
+
+        // don't short circuit! if the stack consists entirely of `Layer`s with
+        // per-layer filters, the `Registry` will return the actual `Interest`
+        // value that's the sum of all the `register_callsite` calls to those
+        // per-layer filters. if we returned an actual `never` interest here, a
+        // `Layered` layer would short-circuit and not allow any `Filtered`
+        // layers below us if _they_ are interested in the callsite.
+        Interest::always()
```

Comment:
it should; however, since i wanted to be able to add per-layer filtering without a `tracing-core` change, that's out of scope for this PR. ideally, we would have a way to extend `Interest` to support caching per-layer filter results when different per-layer filters disagree on a span/event, but we can't do that without core changes.

i'd like to look into that in the future as it should significantly improve performance with per-layer filtering...

Reply:
for sure. i think it's out of scope of this already big PR, but i wanted double-check my understanding.